nodestream_plugin_neptune/neptune_analytics_query_executor.py

```diff
@@ -0,0 +1,163 @@
+
+from aiobotocore.session import get_session
+from nodestream.databases.query_executor import (
+    OperationOnNodeIdentity, OperationOnRelationshipIdentity, QueryExecutor)
+from nodestream.model import (IngestionHook, Node, RelationshipWithNodes,
+                              TimeToLiveConfiguration)
+
+from .ingest_query_builder import NeptuneDBIngestQueryBuilder
+from .query import Query, QueryBatch
+
+
+class NeptuneAnalyticsQueryExecutor(QueryExecutor):
+    def __init__(
+        self,
+        region,
+        graph_id,
+        ingest_query_builder: NeptuneDBIngestQueryBuilder,
+        async_partitions=50,
+    ) -> None:
+        self.session = get_session()
+        self.region = region
+        self.graph_id = graph_id
+        self.ingest_query_builder = ingest_query_builder
+        self.logger = getLogger(self.__class__.__name__)
+        self.async_partitions = async_partitions
+
+    async def upsert_nodes_in_bulk_with_same_operation(
+        self, operation: OperationOnNodeIdentity, nodes: Iterable[Node]
+    ):
+        batched_query = (
+            self.ingest_query_builder.generate_batch_update_node_operation_batch(
+                operation, nodes
+            )
+        )
+        await self.execute_batch(batched_query)
+
+    async def upsert_relationships_in_bulk_of_same_operation(
+        self,
+        shape: OperationOnRelationshipIdentity,
+        relationships: Iterable[RelationshipWithNodes],
+    ):
+        queries = (
+            self.ingest_query_builder.generate_batch_update_relationship_query_batch(
+                shape, relationships
+            )
+        )
+        await self.execute_batch(queries)
+
+    async def perform_ttl_op(self, config: TimeToLiveConfiguration):
+        query = self.ingest_query_builder.generate_ttl_query_from_configuration(config)
+        await self.execute(query)
+
+    async def execute_hook(self, hook: IngestionHook):
+        query_string, params = hook.as_cypher_query_and_parameters()
+        await self.execute(Query(query_string, params))
+
+    def _split_parameters(self, parameters: list):
+        """
+        The intention is to split the parameters evenly so we can make multiple
+        async batch inserts to Neptune. However, I did not notice any meaningful
+        performance improvements.
+
+        It could be that we are doing something incorrect, or that the input
+        size tested (1000 nodes, 2000 edges) weren't enough.
+        """
+        params_count = len(parameters)
+        if not self.async_partitions or params_count < self.async_partitions:
+            partition_size = len(parameters)
+        else:
+            partition_size = math.floor(params_count / self.async_partitions)
+
+        """
+        From Dave we found that the sweet spot is around 100 - 200
+        per batch request. Though this is not a hard rule.
+
+        More investigation on performance is needed.
+        """
+        partition_size = 150
```

Comment:
Should it be configurable, with the default value?

Reply:
I would prefer to leave it fixed pending a proper performance evaluation. If we find that there wide range of "best" values depending on workflow, then we should definitely make it configurable. Currently my expectation is that there will be little need for users to tune this if the default is good. Lets get the data first then decide. It is easier to add a useful config later than to try and remove a redundant one.

Reply:
I tend to agree with this. We have similar fixed params in the neo4j connector for the same reason.